### PR TITLE
Check the base address against the element alignment rather than stride

### PR DESCRIPTION
**Pull request by stephentyrone**
Stride is the gap between elements, not the alignment that each element must satisfy (stride need not even be a power of two).

### CODE REVIEW

**Comment by stephentyrone** [comment]
@swift-ci test

**Review by glessard — APPROVED** [review]
It was right in `withMemoryRebound` and somehow this happened. Thanks!

**Comment by glessard** [comment]
@swift-ci please smoke test

File: stdlib/public/core/UnsafeRawBufferPointer.swift.gyb [hunk 1]
```diff
@@ -793,7 +793,7 @@ extension Unsafe${Mutable}RawBufferPointer {
     }  
 
     _debugPrecondition(
-      Int(bitPattern: base) % MemoryLayout<S.Element>.stride == 0,
+      Int(bitPattern: base) & MemoryLayout<S.Element>.alignment-1 == 0,
```
Comment (glessard) [inline stdlib/public/core/UnsafeRawBufferPointer.swift.gyb:796]: Operator precedence strikes again
```suggestion
      Int(bitPattern: base) & (MemoryLayout<S.Element>.alignment-1) == 0,
```

File: stdlib/public/core/UnsafeRawBufferPointer.swift.gyb [hunk 2]
```diff
@@ -853,7 +853,7 @@ extension Unsafe${Mutable}RawBufferPointer {
         return .init(start: nil, count: 0)
       }
       _debugPrecondition(
-        Int(bitPattern: baseAddress) % MemoryLayout<C.Element>.stride == 0,
+        Int(bitPattern: baseAddress) & MemoryLayout<C.Element>.alignment-1 == 0,
```
Comment (glessard) [inline stdlib/public/core/UnsafeRawBufferPointer.swift.gyb:856]:
```suggestion
        Int(bitPattern: baseAddress) & (MemoryLayout<C.Element>.alignment-1) == 0,
```

File: stdlib/public/core/UnsafeRawBufferPointer.swift.gyb [hunk 3]
```diff
@@ -878,7 +878,7 @@ extension Unsafe${Mutable}RawBufferPointer {
     }
     _internalInvariant(_end != nil)
     _debugPrecondition(
-      Int(bitPattern: baseAddress) % MemoryLayout<C.Element>.stride == 0,
+      Int(bitPattern: baseAddress) & MemoryLayout<C.Element>.alignment-1 == 0,
```
Comment (glessard) [inline stdlib/public/core/UnsafeRawBufferPointer.swift.gyb:881]:
```suggestion
      Int(bitPattern: baseAddress) & (MemoryLayout<C.Element>.alignment-1) == 0,
```

File: stdlib/public/core/UnsafeRawBufferPointer.swift.gyb [hunk 4]
```diff
@@ -940,7 +940,7 @@ extension Unsafe${Mutable}RawBufferPointer {
       return .init(start: nil, count: 0)
     }
     _debugPrecondition(
-      Int(bitPattern: baseAddress) % MemoryLayout<T>.stride == 0,
+      Int(bitPattern: baseAddress) & MemoryLayout<T>.alignment-1 == 0,
```
Comment (glessard) [inline stdlib/public/core/UnsafeRawBufferPointer.swift.gyb:943]:
```suggestion
      Int(bitPattern: baseAddress) & (MemoryLayout<T>.alignment-1) == 0,
```

**Review by glessard — APPROVED** [review]

**Comment by glessard** [comment]
@swift-ci please test

**Comment by stephentyrone** [comment]
Of all the little things about operator precedence in Swift, this is the one I dislike the most.

**Comment by glessard** [comment]
@swift-ci please test

**Comment by glessard** [comment]
@swift-ci please test windows platform

**Comment by stephentyrone** [comment]
@swift-ci smoke test

**Comment by stephentyrone** [comment]
(Using `UInt` for alignment was never a problem on 32b platforms because all the platforms we support have a minimum dynamic allocation alignment of 8B, but we should fix it anyway.)

**Pull request merged**